util/types/json/serdes.go

```diff
@@ -0,0 +1,412 @@
+
+// Serialize means serialize itself into bytes.
+func Serialize(j JSON) []byte {
+	var buffer = new(bytes.Buffer)
+	buffer.WriteByte(j.getTypeCode())
+	j.encode(buffer)
+	return buffer.Bytes()
+}
+
+// Deserialize means deserialize a json from bytes.
+func Deserialize(data []byte) (j JSON, err error) {
+	j = jsonFromTypeCode(data[0])
+	err = jsonDeserFromJSON(j).decode(data[1:])
+	return
+}
+
+var (
+	_ jsonDeser = new(jsonLiteral)
+	_ jsonDeser = new(jsonString)
+	_ jsonDeser = new(jsonArray)
+	_ jsonDeser = new(jsonObject)
+	_ jsonDeser = new(jsonDouble)
+)
+
+const (
+	typeCodeObject  byte = 0x01
+	typeCodeArray   byte = 0x03
+	typeCodeLiteral byte = 0x04
+	typeCodeDouble  byte = 0x0b
+	typeCodeString  byte = 0x0c
+)
+
+const (
+	jsonLiteralNil   = jsonLiteral(0x00)
+	jsonLiteralTrue  = jsonLiteral(0x01)
+	jsonLiteralFalse = jsonLiteral(0x02)
+)
+
+type jsonLiteral byte
+type jsonString string
+type jsonObject map[string]JSON
+type jsonArray []JSON
+type jsonDouble float64
+
+// jsonDeser is for deserialize json from bytes.
+type jsonDeser interface {
+	JSON
+	decode([]byte) error
+}
+
+func jsonDeserFromJSON(j JSON) jsonDeser {
+	if reflect.TypeOf(j).Kind() == reflect.Ptr {
+		if jdeser, ok := j.(jsonDeser); ok {
+			return jdeser
+		}
+	}
+	return nil
+}
+
+func (b jsonLiteral) getTypeCode() byte {
+	return typeCodeLiteral
+}
+
+func (b jsonLiteral) encode(buffer *bytes.Buffer) {
+	buffer.WriteByte(byte(b))
+}
+
+func (b *jsonLiteral) decode(data []byte) error {
+	var bb = (*byte)(unsafe.Pointer(b))
+	*bb = data[0]
+	return nil
+}
+
+func (f jsonDouble) getTypeCode() byte {
+	return typeCodeDouble
+}
+
+func (f jsonDouble) encode(buffer *bytes.Buffer) {
+	binary.Write(buffer, binary.LittleEndian, f)
+}
+
+func (f *jsonDouble) decode(data []byte) error {
+	var reader = bytes.NewReader(data)
+	return binary.Read(reader, binary.LittleEndian, f)
+}
+
+func (s jsonString) getTypeCode() byte {
+	return typeCodeString
+}
+
+func (s jsonString) encode(buffer *bytes.Buffer) {
+	var ss = string(s)
+	var varIntBuf = make([]byte, 9)
+	var varIntLen = binary.PutUvarint(varIntBuf, uint64(len(hack.Slice(ss))))
+	buffer.Write(varIntBuf[0:varIntLen])
+	buffer.Write(hack.Slice(ss))
+}
+
+func (s *jsonString) decode(data []byte) error {
+	var reader = bytes.NewReader(data)
+	length, err := binary.ReadUvarint(reader)
+	if err == nil {
+		var buf = make([]byte, length)
+		_, err = reader.Read(buf)
+		if err == nil {
+			*s = jsonString(string(buf))
+		}
+	}
+	return err
+}
+
+func (m jsonObject) getTypeCode() byte {
+	return typeCodeObject
+}
+
+func (m jsonObject) encode(buffer *bytes.Buffer) {
+	// object ::= element-count size key-entry* value-entry* key* value*
+	var countAndSize = make([]uint32, 2)
+	var countAndSizeLen = len(countAndSize) * 4
+	var keySlice = keysInMap(m)
+
+	var keyEntrysLen = (4 + 2) * len(m)
+	var valueEntrysLen = (1 + 4) * len(m)
+	var keyEntrys = new(bytes.Buffer)
+	var valueEntrys = new(bytes.Buffer)
+	var keys = new(bytes.Buffer)
```

Comment:
I pefer short names, because keyBuffer need keyEntrys and keyEntrysLen respectively, but they are too long.